controllers/nodes/nodes_controller.go

```diff
@@ -0,0 +1,124 @@
+package nodes
+
+import (
+	"context"
+	hcov1beta1 "github.com/kubevirt/hyperconverged-cluster-operator/api/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	operatorhandler "github.com/operator-framework/operator-lib/handler"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hcoutil "github.com/kubevirt/hyperconverged-cluster-operator/pkg/util"
+)
+
+var (
+	log = logf.Log.WithName("controller_nodes")
+)
+
+// RegisterReconciler creates a new Nodes Reconciler and registers it into manager.
+func RegisterReconciler(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	r := &ReconcileNodeCounter{
+		client: mgr.GetClient(),
+	}
+
+	return r
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("nodes-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to the cluster's nodes
+	err = c.Watch(
+		source.Kind(
+			mgr.GetCache(), client.Object(&corev1.Node{}),
+			&operatorhandler.InstrumentedEnqueueRequestForObject[client.Object]{},
+			nodeCountChangePredicate{},
+		))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Custom predicate to detect changes in node count
+type nodeCountChangePredicate struct {
+	predicate.Funcs
+}
+
+func (nodeCountChangePredicate) Update(e event.UpdateEvent) bool {
+	return false
+}
+
+func (nodeCountChangePredicate) Create(e event.CreateEvent) bool {
+	// node is added
+	return true
+}
+
+func (nodeCountChangePredicate) Delete(e event.DeleteEvent) bool {
+	// node is removed
+	return true
+}
+
+func (nodeCountChangePredicate) Generic(e event.GenericEvent) bool {
+	return false
+}
+
+// ReconcileNodeCounter reconciles the nodes count
+type ReconcileNodeCounter struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client              client.Client
+	HyperConvergedQueue workqueue.TypedRateLimitingInterface[reconcile.Request]
+}
+
+// Reconcile updates the nodes count on ClusterInfo singleton
+func (r *ReconcileNodeCounter) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	log.Info("Triggered by node count change")
+	logger := logf.Log.WithName("nodes-controller")
+	clusterInfo := hcoutil.GetClusterInfo()
```

Comment:
If we're reading and writing from several go-routines, then we must protect the data. 

But clusterInfo interface is too large for that, and will force us to protect each and every read method. So I think we must take out `IsControlPlaneHighlyAvailable` and `IsInfrastructureHighlyAvailable` out of the clusterInfo interface, together with their related data, into a new type/interface, and protect them.